hal/src/sercom/v2/uart/config.rs

```diff
@@ -0,0 +1,480 @@
+pub type Clock = pac::MCLK;
+
+impl<P: ValidPads> Config<P> {
+    /// Create a new [`Config`] in the default configuration
+    ///
+    /// This function will enable the corresponding APB clock, reset the
+    /// [`Sercom`] peripheral, and return a [`Config`] in the default
+    /// configuration:
+    ///
+    /// * [`EightBit`] [`CharSize`]
+    /// * No parity
+    /// * One stop bit
+    /// * LSB-first
+    ///
+    /// [`Config`] takes ownership of the [`Sercom`] and [`Pads`](super::Pads).
+    ///
+    /// Users must configure GCLK manually. The `freq` parameter represents the
+    /// GCLK frequency for this [`Sercom`] instance.
+    #[inline]
+    pub fn new(clk: &Clock, mut sercom: P::Sercom, pads: P, freq: impl Into<Hertz>) -> Self {
+        sercom.enable_apb_clock(clk);
+        Self::default(sercom, pads, freq).bit_order(BitOrder::LsbFirst)
+    }
+
+    /// Create a new [`Config`] in the default configuration
+    #[inline]
+    fn default(sercom: P::Sercom, pads: P, freq: impl Into<Hertz>) -> Self {
+        let mut registers = Registers::new(sercom);
+        registers.swrst();
+
+        // Enable internal clock mode
+        registers.configure_mode();
+        registers.configure_pads(P::RXPO as u8, P::TXPO as u8);
+        registers.set_char_size(EightBit::BITS);
+
+        Self {
+            registers,
+            pads,
+            chsize: PhantomData,
+            freq: freq.into(),
+        }
+    }
+}
+
+impl<P, C> Config<P, C>
+where
+    P: ValidPads,
+    C: CharSize,
+{
+    /// Change the [`Config`] [`CharSize`]
+    #[inline]
+    fn change<C2>(self) -> Config<P, C2>
+    where
+        C2: CharSize,
+    {
+        Config {
+            registers: self.registers,
+            pads: self.pads,
+            chsize: PhantomData,
+            freq: self.freq,
+        }
+    }
+
+    /// Trigger the [`Sercom`]'s SWRST and return a [`Config`] in the
+    /// default configuration.
+    #[inline]
+    pub fn reset(self) -> Config<P> {
+        Config::default(self.registers.free(), self.pads, self.freq)
+    }
+
+    /// Consume the [`Config`], reset the peripheral, and return the [`Sercom`]
+    /// and [`Pads`](super::Pads)
+    #[inline]
+    pub fn free(mut self) -> (P::Sercom, P) {
+        self.registers.swrst();
+        (self.registers.free(), self.pads)
+    }
+
+    /// Change the [`CharSize`].
+    #[inline]
+    pub fn char_size<C2: FixedCharSize>(mut self) -> Config<P, C2> {
+        self.registers.set_char_size(C2::BITS);
+        self.change()
+    }
+
+    /// Change the [`CharSize`] to [`DynCharSize`]. The UART's character
+    /// size will be changed to the provided `C2`, which is a [`FixedCharSize`]
+    /// type, and can then be changed dynamically on an enabled [`Uart`]
+    /// without changing the underlying [`Config`]'s type through the
+    /// [`reconfigure`](Uart::reconfigure) method.
+    #[inline]
+    pub fn dyn_char_size<C2: FixedCharSize>(mut self) -> Config<P, DynCharSize> {
```

Comment:
Hmm. I still find this a bit strange. In one function, you both set the `Config` struct to `DynCharSize` mode and also set the actual character size. To me, it would make more sense for the `dyn_char_size` function to change it to `DynCharSize` mode *only*, leaving it with the default character size (`CharSizeEnum::EightBit`). Then a separate call to `set_dyn_char_size` would actually set the desired character size. That way, if you're using `DynCharSize`, you only need to interact with `CharSizeEnum` and never the `FixedCharSize` types.

Thoughts?

Reply:
Yeah I think that makes perfect sense.